### PR TITLE
Analytical gradient REPS and C-REPS

### LINKED ISSUE #70: REPS / C-REPS dual function gradient
When minimizing the dual function in REPS / C-REPS, why is the gradient approximated numerically rather than using the analytical expressions provided [here](https://www.sciencedirect.com/science/article/pii/S0004370214001374)? Is this for numerical stability?

### PR DESCRIPTION
**Pull request by RicardoDominguez**
Fixes #70 

For the formulas used refer to [this](https://www.sciencedirect.com/science/article/pii/S0004370214001374).

### CODE REVIEW

File: bolero/optimizer/reps.py [hunk 1]
```diff
@@ -49,15 +49,21 @@ def solve_dual_reps(R, epsilon, min_eta):
...
+        return f, np.array([d_eta])
 
     # Lower bound for Lagrangian eta
     bounds = np.array([[min_eta, None]])
     # Start point of optimization
     x0 = [1]
 
     # Perform the actual optimization of the dual function
-    r = fmin_l_bfgs_b(g, x0, approx_grad=True, bounds=bounds)
+    r = fmin_l_bfgs_b(g, x0, approx_grad=None, bounds=bounds)
```
Comment (AlexanderFabisch) [inline bolero/optimizer/reps.py]: you can remove the keyword argument `approx_grad`

File: bolero/optimizer/creps.py [hunk 2]
```diff
@@ -53,9 +53,18 @@ def solve_dual_contextual_reps(S, R, epsilon, min_eta):
...
+
+        s_mean = S.mean(axis=0)
+        R_over_eta = (R - nu.dot(S.T)) / eta
+        log_sum_exp = logsumexp(R_over_eta, b=1.0 / n_samples_per_update)
+        Z = np.exp(R_over_eta - R_over_eta.max())
+        Z_sum = Z.sum()
+
+        f = eta * (epsilon + log_sum_exp) + nu.T.dot(s_mean)
+        d_eta = epsilon + log_sum_exp - (Z.dot(R_over_eta) / Z_sum)
+        d_nu = s_mean - (Z.dot(S) / Z_sum)
```
Comment (AlexanderFabisch) [inline bolero/optimizer/creps.py]: You can remove parentheses here.

File: bolero/optimizer/creps.py [hunk 3]
```diff
@@ -53,9 +53,18 @@ def solve_dual_contextual_reps(S, R, epsilon, min_eta):
...
-                                b=1.0 / n_samples_per_update))
+
+        s_mean = S.mean(axis=0)
+        R_over_eta = (R - nu.dot(S.T)) / eta
+        log_sum_exp = logsumexp(R_over_eta, b=1.0 / n_samples_per_update)
+        Z = np.exp(R_over_eta - R_over_eta.max())
+        Z_sum = Z.sum()
+
+        f = eta * (epsilon + log_sum_exp) + nu.T.dot(s_mean)
+        d_eta = epsilon + log_sum_exp - (Z.dot(R_over_eta) / Z_sum)
```
Comment (AlexanderFabisch) [inline bolero/optimizer/creps.py]: You can remove parentheses here.

File: bolero/optimizer/creps.py [hunk 4]
```diff
@@ -53,9 +53,18 @@ def solve_dual_contextual_reps(S, R, epsilon, min_eta):
     def g(x):  # Objective function
         eta = x[0]
         nu = x[1:]
-        return (eta * epsilon + nu.T.dot(S.mean(axis=0)) +
-                eta * logsumexp((R - nu.dot(S.T)) / eta,
-                                b=1.0 / n_samples_per_update))
+
```
Comment (AlexanderFabisch) [inline bolero/optimizer/creps.py]: I tried to improve the efficiency of the code but noticed that you did a pretty good job already. :)

File: bolero/optimizer/creps.py [hunk 5]
```diff
@@ -64,7 +73,8 @@ def g(x):  # Objective function
 
     # Perform the actual optimization of the dual function
     #r = NLP(g, x0, lb=lb).solve('ralg', iprint=-10)
-    r = fmin_l_bfgs_b(g, x0, approx_grad=True, bounds=bounds)
+    r = fmin_l_bfgs_b(g, x0, approx_grad=None, bounds=bounds)
```
Comment (AlexanderFabisch) [inline bolero/optimizer/creps.py]: you can remove the keyword argument `approx_grad`

**Comment by RicardoDominguez** [comment]
I addressed the changes you suggested.

**Comment by AlexanderFabisch** [comment]
Merged to master. Thanks for your contribution!

**Pull request closed without merging**